Allow other users to use security keys

benthorner opened the pull request:
https://www.pivotaltracker.com/story/show/178609960

This relaxes some of the restrictions around security keys and uses a new API
flag [1] to also let non-platform admins use the feature.

[1]: https://github.com/alphagov/notifications-api/pull/3280

CrystalPea reviewed (commented):
Hey, I looked through the code, it looks good, some comments and test names could be a little more descriptive.

I also tested it on my local machine and it works fine 👍🏼

benthorner commented on app/main/views/webauthn_credentials.py:
```diff
@@ -105,12 +101,6 @@ def webauthn_complete_authentication():
     user_id = session['user_details']['id']
     user_to_login = User.from_id(user_id)
 
-    if not user_to_login.webauthn_auth:
```
I explained the reason for removing this in the second commit: https://github.com/alphagov/notifications-admin/pull/3947/commits/ec53fa2f3d2c52e2b1bc042638de2379268dc394. Another reason to add is that part of the rollout plan for WebAuthn would also remove both of these checks: any user will be able to choose how they log in.

benthorner commented on app/main/forms.py:
```diff
@@ -1009,14 +1009,10 @@ def from_user(cls, user, service_id, **kwargs):
-        if user.platform_admin:
+
+        # If a user logs in with a security key, we generally don't want a service admin to be able to change this.
+        # As well as enforcing this in the backend, we need to delete the field to prevent validation errors.
```
So here a "field" refers to a field in the form: all forms are composed of fields that a user fills in. In this case, we're deleting the `login_authentication` field, which is a bunch of radio buttons. Deleting the field means it doesn't appear on the form, or take part in any later processing.

Does that help?

CrystalPea replied:
yes, that helps a lot,  thanks! Could we maybe change the comment to: 
`As well as enforcing this in the backend, we need to hide the authentication method options to prevent validation errors.`? I feel that would be more explicit.

benthorner replied:
Hmm...I think "delete" is more accurate here: we're not hiding the option, we really are deleting it entirely from the form, so it's not visible, but also not part of the processing. Hiding it suggests it's just a visual thing.

I'd prefer to keep it as-is for now, and let someone else change it if they feel the need. My hope is that we'll do enough of the WebAuthn work that we can delete this feature permanently...

CrystalPea replied:
ok, so how about 
```suggestion
        # As well as enforcing this in the backend, we need to delete the field used to set authentication method to prevent validation errors.
```

benthorner replied:
I've amended the commit to [something a bit different](https://github.com/alphagov/notifications-admin/pull/3947/commits/a1b4ccc246faf6ddd96da7bfb7c39352b7e6c94c#diff-a1c8d24b22d4478fe71f75fd43b71b18dd82aae97bc63de84473a6da1902909bR1014). Might help if you explain what the problem is, as I'm not sure how these changes add value.

CrystalPea replied:
ok, that is more clear. It's just that `field` was so unclear to me, I wouldn't know what that sentence means if I read it, and supposedly it is there to provide context, so it should be understood by people.

benthorner commented on app/main/views/manage_users.py:
```diff
@@ -142,9 +142,9 @@ def edit_user_permissions(service_id, user_id):
-        # we avoid changing the auth type to prevent it being switched to something less secure
-        if service_has_email_auth and not user.platform_admin:
+        # Only change the auth type if this is supported for a service. If a user logs in with a
+        # security key, we generally don't want service admins to be able to change this.
```
Sort of. It's not accurate because sometimes it _is_ OK for a broadcast service user to switch to something less secure, but we can't say exactly when at the moment, hence the blanket restriction.

I've replaced it with the following:

```
# Only change the auth type if this is supported for a service. If a user logs in with a
# security key, we generally don't want them to be able to use something less secure.
```

benthorner commented on tests/app/main/views/test_user_profile.py:
```diff
@@ -358,7 +358,13 @@ def test_can_reenable_platform_admin(client_request, platform_admin_user):
 
 
-def test_normal_user_doesnt_see_security_keys(client_request):
+def test_user_doesnt_see_security_keys_unless_can_use_webauthn(
```
Sure, that's fair: I was being a little terse. 

I've amended the commit for this and the other test.

benthorner commented:
Thanks for the review @CrystalPea.

I think I've addressed all your comments, but @leohemsted might want to take a second look at https://github.com/alphagov/notifications-admin/pull/3947#discussion_r662443003.

benthorner commented on tests/app/main/views/test_manage_users.py:
```diff
@@ -742,7 +742,6 @@ def test_cant_edit_non_member_user_permissions(
     assert mock_set_user_permissions.called is False
 
 
-@pytest.mark.parametrize('current_auth_type', ['webauthn_auth', 'email_auth'])
```
The test will now fail with `webauthn_auth`, since we now prevent all users from switching away from this auth type (as opposed to just platform admins).

Does that make sense?

CrystalPea replied:
yeah, it does, thanks a lot!
